### PR TITLE
Custom audit policy

### PR DESCRIPTION
**Pull request by vpnachev**
**What this PR does / why we need it**: With this PR the end user could specify the [audit policy](https://kubernetes.io/docs/tasks/debug-application-clustasder/audit/#audit-policy) for the `kube-apiserver` of the `shoot` cluster.

**Which issue(s) this PR fixes**:
Fixes: None

**Special notes for your reviewer**:
None

**Release note**:

```noteworthy user
Users can now define custom audit policies for the kube-apiserver of their shoot clusters. Please find a detailed guide how to use this feature [here](https://github.com/gardener/gardener/tree/master/docs/usage/shoot_auditpolicy.md).
```

### CODE REVIEW

**Comment by rfranzke** [comment]
Please comment here as soon as this PR is no longer "WIP".

**Comment by vpnachev** [comment]
Hi @rfranzke. This PR is ready for review.
I have switched the `LocalObjectRef` to `ObjectRef`, also yesterday I rebased on master.

File: pkg/apis/garden/types.go [hunk 1]
```diff
@@ -1202,6 +1202,25 @@ type KubeAPIServerConfig struct {
...
+type AuditConfig struct {
+	// AuditPolicy contains audit policy for kube-apiserver
+	// +optional
+	AuditPolicy *AuditPolicy
+}
+
+// AuditPolicy contains audit policy for kube-apiserver
+type AuditPolicy struct {
+	// +optional
+	// AuditPolicy *auditv1beta1.Policy
```
Comment (rfranzke) [inline pkg/apis/garden/types.go]: Can we remove this as `ConfigMapRef` is the only option ATM?

File: pkg/apis/garden/v1beta1/types.go [hunk 2]
```diff
@@ -1192,6 +1192,25 @@ type KubeAPIServerConfig struct {
...
+type AuditConfig struct {
+	// AuditPolicy contains audit policy for kube-apiserver
+	// +optional
+	AuditPolicy *AuditPolicy `json:"auditPolicy,omitempty"`
+}
+
+// AuditPolicy contains audit policy for kube-apiserver
+type AuditPolicy struct {
+	// +optional
+	// AuditPolicy *auditv1beta1.Policy
```
Comment (rfranzke) [inline pkg/apis/garden/v1beta1/types.go]: Can we remove this as `ConfigMapRef` is the only option ATM?

File: pkg/apis/garden/validation/validation_test.go [hunk 3]
```diff
@@ -4457,6 +4465,49 @@ var _ = Describe("validation", func() {
...
+				}))
+			})
+
+			It("should forbid nil configMapRef", func() {
+				shoot.Spec.Kubernetes.KubeAPIServer.AuditConfig.AuditPolicy.ConfigMapRef = nil
+				errorList := ValidateShoot(shoot)
+
+				Expect(len(errorList)).To(Equal(1))
+				Expect(*errorList[0]).To(MatchFields(IgnoreExtras, Fields{
+					"Type":  Equal(field.ErrorTypeInvalid),
```
Comment (rfranzke) [inline pkg/apis/garden/validation/validation_test.go]: Shouldn't this be `ErrorTypeRequired`?

File: pkg/apis/garden/validation/validation_test.go [hunk 4]
```diff
@@ -4457,6 +4465,49 @@ var _ = Describe("validation", func() {
...
+				}))
+			})
+
+			It("should forbid nil auditPolicy", func() {
+				shoot.Spec.Kubernetes.KubeAPIServer.AuditConfig.AuditPolicy = nil
+				errorList := ValidateShoot(shoot)
+
+				Expect(len(errorList)).To(Equal(1))
+				Expect(*errorList[0]).To(MatchFields(IgnoreExtras, Fields{
+					"Type":  Equal(field.ErrorTypeInvalid),
```
Comment (rfranzke) [inline pkg/apis/garden/validation/validation_test.go]: Shouldn't this be `ErrorTypeRequired`?

File: pkg/operation/hybridbotanist/controlplane.go [hunk 5]
```diff
@@ -264,6 +265,31 @@ func (b *HybridBotanist) DeployKubeAPIServer() error {
 				admissionPlugins = append(admissionPlugins, plugin)
 			}
 		}
+
+		if apiServerConfig.AuditConfig != nil &&
+			apiServerConfig.AuditConfig.AuditPolicy != nil {
```
Comment (rfranzke) [inline pkg/operation/hybridbotanist/controlplane.go]: Please keep the `if` condition in the same line.

File: pkg/operation/hybridbotanist/controlplane.go [hunk 6]
```diff
@@ -264,6 +265,31 @@ func (b *HybridBotanist) DeployKubeAPIServer() error {
...
+
+			cmRefName := apiServerConfig.AuditConfig.AuditPolicy.ConfigMapRef.Name
+			cmRefNamespace := apiServerConfig.AuditConfig.AuditPolicy.ConfigMapRef.Namespace
+			if len(cmRefNamespace) == 0 {
+				cmRefNamespace = b.Shoot.Info.Namespace
+			}
+			auditPolicyCm, err := b.K8sGardenClient.Clientset().CoreV1().ConfigMaps(cmRefNamespace).Get(cmRefName, metav1.GetOptions{})
```
Comment (rfranzke) [inline pkg/operation/hybridbotanist/controlplane.go]: Please use `b.K8sGardenClient.GetConfigMap(cmRefNamespace, cmRefName)` (https://github.com/gardener/gardener/blob/master/pkg/client/kubernetes/types.go#L83).

File: pkg/apis/garden/types.go [hunk 7]
```diff
@@ -1202,6 +1202,25 @@ type KubeAPIServerConfig struct {
...
+
+	// +optional
+	ConfigMapRef *corev1.ObjectReference
```
Comment (rfranzke) [inline pkg/apis/garden/types.go]: This should be a `LocalObjectReference`  to prevent referencing ConfigMaps from other namespaces we might probably have no permissions for.

File: pkg/apis/garden/v1beta1/types.go [hunk 8]
```diff
@@ -1192,6 +1192,25 @@ type KubeAPIServerConfig struct {
...
+
+	// +optional
+	ConfigMapRef *corev1.ObjectReference `json:"configMapRef,omitempty"`
```
Comment (rfranzke) [inline pkg/apis/garden/v1beta1/types.go]: This should be a `LocalObjectReference`  to prevent referencing ConfigMaps from other namespaces we might probably have no permissions for.

**Review by rfranzke — CHANGES_REQUESTED** [review]
Thanks, looks good, just a few suggestions.
Apart from the things inline:

1. Please add an example configuration to the `.spec.kubernetes.kubeAPIServer` section in the example Shoot manifests: https://github.com/gardener/gardener/blob/master/hack/templates/resources/90-shoot.yaml.tpl#L212
1. Could you please also add a small `shoot_auditpolicy.yaml` markdown document [here](https://github.com/gardener/gardener/tree/master/docs/usage) and quickly describe the behaviour and how to use it (that a missing audit policy reference in the Shoot defaults to a "standard" audit policy; that the referenced configmap needs a key `policy` (why not `policy.yaml`); etc.).
1. We do not re-queue the Shoots referencing configmap `x` when `x` updates, i.e., audit policy updates may take hours to get applied. Was that a deliberate decision?

Also, please rebase to the latest `master` and squash your commits.

File: pkg/apis/garden/validation/validation.go [hunk 9]
```diff
@@ -1524,6 +1524,21 @@ func validateKubernetes(kubernetes garden.Kubernetes, fldPath *field.Path) field
 				allErrs = append(allErrs, field.Required(idxPath.Child("name"), "must provide a name"))
 			}
 		}
+
+		if auditConfig := kubeAPIServer.AuditConfig; auditConfig != nil {
```
Comment (adracus) [inline pkg/apis/garden/validation/validation.go:1564]: Can you please move this to an own function, e.g. `ValidateAuditConfig`? Makes testing + reasoning easier.

File: pkg/apis/garden/validation/validation.go [hunk 10]
```diff
@@ -1524,6 +1524,21 @@ func validateKubernetes(kubernetes garden.Kubernetes, fldPath *field.Path) field
...
+			auditPath := fldPath.Child("kubeAPIServer", "auditConfig")
+
+			if auditPolicy := auditConfig.AuditPolicy; auditPolicy != nil {
```
Comment (adracus) [inline pkg/apis/garden/validation/validation.go]: Same here, move to own function.

**Review by adracus — DISMISSED** [review]

File: pkg/apis/garden/types.go (hunk 7 above)
Comment (vpnachev) [reply pkg/apis/garden/types.go]: I have talked about this with @vlerenc. I switched from `LocalObjectReference` to `ObjectReference` because of several reasons:
* the Garden operator may define cluster-wide audit policies in some common namespace (`garden` for example). In this way the operator does not have to replicate the audit policies across all namespaces and support multiple replicas of common resource.
* the Garden operator may disallow end users to deploy their audit policies (ConfigMaps) in the namespace of their Project. In this way the end users will be obligated to use only predefined audit policies.

The `ObjectReference` has its own downsides:
* a shoot from project `X` may use audit policy from project `Y`

As a solution, I could limit the `ObjectReference` to refer config maps in the same namespace where the shoot resides, or in the common `garden` namespace.

File: pkg/apis/garden/types.go (hunk 1 above)
Comment (vpnachev) [reply pkg/apis/garden/types.go]: Done.

File: pkg/operation/hybridbotanist/controlplane.go (hunk 5 above)
Comment (vpnachev) [reply pkg/operation/hybridbotanist/controlplane.go]: Done

File: pkg/operation/hybridbotanist/controlplane.go (hunk 6 above)
Comment (vpnachev) [reply pkg/operation/hybridbotanist/controlplane.go]: Done

**Review by vpnachev — COMMENTED** [review]
I have implemented the proposed changes.
@rfranzke  I have added my opinion why I switched from `LocalObjectReference`. I will squash the commits just right after the change is approved. 
I plan to implement additional controller, which will watch for changes in these configMaps and it will re-queue the shoots which reference them. But this will be provided later with other PR.


@Adracus now the part related to Audit validation is as clear and readable as the one for Admission plugins. What do you think?

**Comment by rfranzke** [comment]
Thanks you very much!

1. Feedback regarding `LocalObjectReference` vs. `ObjectReference`:

> the Garden operator may define cluster-wide audit policies in some common namespace (garden for example). In this way the operator does not have to replicate the audit policies across all namespaces and support multiple replicas of common resource.

Hm, if you say cluster-wide then do you mean that every Shoot that does not have a custom audit policy shall get it? If so then we should allow configuration of the default audit policy available in the configuration of the Gardener (right now, we "hard-code" it in the Docker image/Helm chart values, but we could make this default audit policy configurable from outside).

> the Garden operator may disallow end users to deploy their audit policies (`ConfigMap`s) in the namespace of their Project. In this way the end users will be obligated to use only predefined audit policies.

How would that happen? Restricting the users from creating `ConfigMap`s in their namespace? Why is this related to whether we use `LocalObjectReference` or `ObjectReference`? What happens if the user does not specify an audit policy at all?

> As a solution, I could limit the ObjectReference to refer config maps in the same namespace where the shoot resides, or in the common garden namespace.

I don't understand, if we limit it that way we can go with the `LocalObjectReference` type right away.

Generally, what do we want to achieve? We want to give the users freedom to define their own Shoot specific audit policy. Now you say the Gardener providers might want to define cluster-wide policies which might affect every Shoot (maybe also supporting this only at project level?). In order to achieve this we would have to update the audit policy config of each and every Shoot to reference this `ConfigMap`. Who will do that? And what happens if the user changes the reference back to his own `ConfigMap`? How do we prevent this? If we don't prevent it what is the benefit of having this cluster-wide/project-wide policy?

PS: Maybe we can put such thoughts/discussions which seem to have happened already some weeks ago directly to the respective issue/PR so that we can discuss upfront.

2. Feedback regarding changes to the audit policy `ConfigMap`s:

> I plan to implement additional controller, which will watch for changes in these configMaps and it will re-queue the shoots which reference them. But this will be provided later with other PR.

Should we really bring in a PR that is incomplete? If we support referencing `ConfigMap`s containing custom audit policies then users expect us to update their clusters when the content changes. No additional controller is needed. We just need another event handler on `ConfigMap` update events in the Shoot controller that finds for a given `ConfigMap` `x` all Shoots referencing `x` to re-queue them into the Shoot queue. (Similar to what the `Machine` controller in the machine-controller-manager does: It also watches `Node` objects and re-queues the corresponding `Machine` object on changes).

3. We are not validating that the referenced `ConfigMap` exists. However, we are doing this kind of validation for all other fields which reference external objects. In order to have a common implementation please add a proper validation in the [`ResourceReferenceManager`](https://github.com/gardener/gardener/tree/master/plugin/pkg/global/resourcereferencemanager) admission plugin.

**Comment by vpnachev** [comment]
Thank you @rfranzke for your opinion. 

Let me elaborate on some points.
1. `LocalObjectReference` vs. `ObjectReference`
> Hm, if you say cluster-wide then do you mean that every Shoot that does not have a custom audit policy shall get it? 

No, I mean that several policies will be deployed in the `garden` namespace (or some other well know namespace) and they will be available for usage from all of the shoot clusters. After that end users may be advised to chose one of these audit policies for their shoot clusters, otherwise they have to stay with the default one (see below which is actually the default policy) or they have to investigate resources in defining their own audit policy for their cluster. Yes, the policies provided by the garden operator will not satisfy every end user, but they will be good enough for starting point. 

> If so then we should allow configuration of the default audit policy available in the configuration of the Gardener (right now, we "hard-code" it in the Docker image/Helm chart values, but we could make this default audit policy configurable from outside).

This hard-coded policy in the Docker image of the `kube-apiserver` Helm chart is actually the default policy for every `Shoot` cluster. It is applied if and only if the `AuditPolicy` section of the `Shoot` is omitted. If it does not fit somebody`s use case, he/she could define his/her own audit policy and deploy it in the `Garden` cluster as `ConfigMap` and reference it from the `Shoot` spec. So, I think that we do not have to do anything about this hard-coded policy.

Of course, these policies could be listed in the documentation and whoever wants to use one of them, he/she has to deploy it in the namespace of the `Shoot`. So both options could be addressed with `LocalObjectReference` as well. 

> How would that happen? Restricting the users from creating ConfigMaps in their namespace? Why is this related to whether we use LocalObjectReference or ObjectReference? What happens if the user does not specify an audit policy at all?

With `RBAC` the creation of `ConfigMap` resources in any namespace could be disallowed for the end users and the garden provider take full control to define the audit policies. Yes, this is too tight limitation, but I am not aware of any other use case when the end users need to create config maps in the `Garden` cluster. Am I right ? In this way the end user will be limited to stay with the hard-coded policy or to get one already deployed by garden operator.

> I don't understand, if we limit it that way we can go with the LocalObjectReference type right away.

No, we cannot. Because we cannot find out which namespace is referred - `garden` namespace or the namespace where the `Shoot` resides. `LocalObjectRef` could specify only `Name` of the resource and nothing more. 

> Generally, what do we want to achieve? We want to give the users freedom to define their own Shoot specific audit policy. Now you say the Gardener providers might want to define cluster-wide policies which might affect every Shoot (maybe also supporting this only at project level?).

Our main task is to allow end users to bring their own audit policies. And this will be the default behavior of the `Gardener`.
Some Garden providers may want to ensure that the shoot clusters deployed in their Garden landscape is compliant with their standards for auditing - which means that only verified audit policies could be used for the Shoot clusters. In this case, the Garden provider will provide the policies and will obligate end users to use one of them, which could be achieved with just one more `RBAC`. The Garden provider will disallow end users to manage `ConfigMap` resources, otherwise the end users will not be actually obligated to use one of the provided policies.
The difference between `LocalObjectReference` and `ObjectReference` here is whether the Gardener provider has to replicate these audit policies in every namespace or he could deploy and maintain them easily in single namespace.

> In order to achieve this we would have to update the audit policy config of each and every Shoot to reference this ConfigMap. Who will do that? And what happens if the user changes the reference back to his own ConfigMap? How do we prevent this? If we don't prevent it what is the benefit of having this cluster-wide/project-wide policy?

No body will do it. The garden Docker images could be rebuilt with other hard-coded default policy, which is actually valid audit policy for particular Garden provider. The users could not change the reference to their own `ConfigMap` because they will have no `ConfigMap`(they are not allowed to manage config maps and the garden provider deleted their previously defined config maps). The benefit is to ease the Garden provider in maintenance of up-to-date validated audit policies in its Garden offering, while this will be achieved only with one additional `RBAC` rule disallowing end user to manage config maps.

2. Feedback regarding changes to the audit policy ConfigMaps

You are right and I am aware that this is the desired behavior. I will address it. Thank you for pointing appropriate examples.

3. Validating Referenced Objects
I will do it.

****

#### In summary of `LocalObjectReference` vs. `ObjectReference`

__I think that we have to clear who is responsible to provide and support audit policies:__
* the garden operator will maintain a few policies and has to guarantee that end users does not bring their own non-validated policy
OR
* every end user could brings and maintains his own policy, while the garden provider may or may not give his users a few examples of audit policies with the documentation of its offering

@vlerenc , @mvladev , @shturec  what do you think -  who have to be responsible for the used audit policy - garden provider or end user of a garden offering?

**Comment by rfranzke** [comment]
> No, I mean that several policies will be deployed in the garden namespace (or some other well know namespace) and they will be available for usage from all of the shoot clusters. After that end users may be advised to chose one of these audit policies for their shoot clusters, otherwise they have to stay with the default one (see below which is actually the default policy) or they have to investigate resources in defining their own audit policy for their cluster. Yes, the policies provided by the garden operator will not satisfy every end user, but they will be good enough for starting point.

OK, understood, but if we want to only want to provide kind of "best practices" audit policies to end users I would rather go and document them somewhere instead of going this approach. We anyway have to document it somewhere (because the name and namespace of these configs cannot be guessed by end-users, and they certainly want to see want gets audit logged anyway): We don't win anything if we pre-create them in the garden namespace.

> This hard-coded policy in the Docker image of the kube-apiserver Helm chart is actually the default policy for every Shoot cluster. It is applied if and only if the AuditPolicy section of the Shoot is omitted. If it does not fit somebodys use case, he/she could define his/her own audit policy and deploy it in theGardencluster asConfigMapand reference it from theShoot` spec. So, I think that we do not have to do anything about this hard-coded policy.

That was not the point, I was not addressing the fact that it's "hard-coded", I just wanted to say that if we want the behavior described by you in the last comment then we should think about changing this and making it configurable from the outside.

> With RBAC the creation of ConfigMap resources in any namespace could be disallowed for the end users and the garden provider take full control to define the audit policies. Yes, this is too tight limitation, but I am not aware of any other use case when the end users need to create config maps in the Garden cluster. Am I right ? In this way the end user will be limited to stay with the hard-coded policy or to get one already deployed by garden operator.

Right, at the moment the only use case for creating config maps would be the custom audit policy use case. As you said: Restricting the whole creation of ConfigMaps only because we want to disallow custom audit policies is a very tight restriction and feels wrong.

> Our main task is to allow end users to bring their own audit policies. And this will be the default behavior of the Gardener.
> Some Garden providers may want to ensure that the shoot clusters deployed in their Garden landscape is compliant with their standards for auditing - which means that only verified audit policies could be used for the Shoot clusters. In this case, the Garden provider will provide the policies and will obligate end users to use one of them, which could be achieved with just one more RBAC. The Garden provider will disallow end users to manage ConfigMap resources, otherwise the end users will not be actually obligated to use one of the provided policies.
> The difference between LocalObjectReference and ObjectReference here is whether the Gardener provider has to replicate these audit policies in every namespace or he could deploy and maintain them easily in single namespace.

No, it cannot be achieved with "just disallowing users from managing config maps". Users can easily change their Shoot to not reference any audit policy so that the wanted compliance gets violated.

In the end, the way how it's currently implemented does not fit to what you described/want. I have no problem regarding `ObjectReference`, but if we do it that way we need a cleaner concept of what we want to achieve with it and, more importantly, how we achieve it.

**Comment by rfranzke** [comment]
How do you want to continue now?

**Comment by vpnachev** [comment]
I switched back to the `LocalObjectReference`. Also I have added validation for the referenced resources. 
Tomorrow I will go on with the config map event handler.

At the end, the shoot owner will have the full control of the audit policy, while Garden provider could recommend audit policies via documentation of its offering. 

I will ping you for another round of code review when I am ready.

**Comment by rfranzke** [comment]
OK, thank you very much!

**Comment by rfranzke** [comment]
FYI (from the Kubernetes 1.12 release notes): The audit.k8s.io api group has been upgraded from v1beta1 to v1.

**Comment by mvladev** [comment]
> We are not validating that the referenced ConfigMap exists. However, we are doing this kind of validation for all other fields which reference external objects. In order to have a common implementation please add a proper validation in the ResourceReferenceManager admission plugin.

@rfranzke this is a big anti-pattern in K8S - it forces you to have ordering for creation of resources.
If a `Pod` requires a `ConfigMap` reference, you can create the `Pod` first and then the `ConfigMap` - for the `kube-apiserver` the order of creation doesn't matter. It's up to the individual controller (in this case `kubelet`) to report the status of the system.

The only related change in terms of admission is to check if the `end-user` can actually `get` said `ConfigMap`

**Comment by rfranzke** [comment]
Agreed, might be, but that's how it is implemented currently. We (you and I) had discussed this months ago and we knew that this behaviour differs from Kubernetes conventions, however, we decided to keep it for the time being. As we don't want to change this now but start working on extensibility soon I suggest to adapt the implementation to the current behaviour. We can definitely have a follow-up discussion about whether we should remove this behaviour in the future. OK?

File: charts/seed-controlplane/charts/kube-apiserver/templates/_helpers.tpl [hunk 11]
```diff
@@ -69,3 +69,49 @@
...
 {{- end }}
 {{- end -}}
+
+{{- define "kube-apiserver.auditversion" -}}
+{{- if semverCompare ">= 1.12" .Values.kubernetesVersion -}}
+audit.k8s.io/v1
+{{- else -}}
+audit.k8s.io/v1beta1
+{{- end -}}
+{{- end -}}
```
Comment (rfranzke) [inline charts/seed-controlplane/charts/kube-apiserver/templates/_helpers.tpl:79]: If you want to keep that section in this file then please remove it from [here](https://github.com/gardener/gardener/blob/master/charts/api-versions/templates/_versions.tpl#L21-L27) (or vice versa).

File: hack/templates/resources/90-shoot.yaml.tpl [hunk 12]
```diff
@@ -237,6 +237,9 @@ spec:
   #         clusterDefaultNodeSelector: <node-selectors-labels>
   #         namespace1: <node-selectors-labels>
   #         namespace2: <node-selectors-labels>
+  #   auditConfig:
+  #     auditPolicy:
+  #       name: auditpolicy
```
Comment (rfranzke) [inline hack/templates/resources/90-shoot.yaml.tpl]: It's not clear without further documentation that `name` references a `ConfigMap`. Can we rename this field to `configMapName` or `configMapRef: {name: <name>}`?

File: example/95-audit-policy-for-shoot.yaml [hunk 13]
```diff
@@ -0,0 +1,12 @@
+apiVersion: v1
```
Comment (rfranzke) [inline example/95-audit-policy-for-shoot.yaml]: All example file names start with the `kind`, hence, please rename this file to e.g. `95-configmap-custom-audit-policy.yaml`.

File: example/95-audit-policy-for-shoot.yaml [hunk 14]
```diff
@@ -0,0 +1,12 @@
...
+kind: ConfigMap
+metadata:
+  name: auditpolicy
```
Comment (rfranzke) [inline example/95-audit-policy-for-shoot.yaml]: In the shoot example manifests we state [`namespace: garden-dev`](https://github.com/gardener/gardener/blob/master/example/90-shoot-aws.yaml#L6), thus, to make it more clear please also state it here.

File: pkg/apis/garden/validation/validation.go (hunk 9 above)
Comment (rfranzke) [reply pkg/apis/garden/validation/validation.go:1564]: I've no strong opinion regarding this one as the other parts are also not executed in separate functions (oidc, admission plugins, etc.).

File: pkg/controller/factory.go [hunk 15]
```diff
@@ -42,6 +42,10 @@ import (
 	"k8s.io/client-go/tools/record"
 )
 
+const (
+	configMapConcurentSyncs = 5
```
Comment (rfranzke) [inline pkg/controller/factory.go]: typo

File: pkg/controller/shoot/shoot_control.go [hunk 16]
```diff
@@ -32,13 +32,37 @@ import (
...
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/apimachinery/pkg/util/sets"
 	"k8s.io/client-go/tools/cache"
 	"k8s.io/client-go/tools/record"
 )
 
+func (c *Controller) configMapAdd(obj interface{}) {
```
Comment (rfranzke) [inline pkg/controller/shoot/shoot_control.go]: Please move the `ConfigMap` parts introduced here in their own file `pkg/controller/shoot/configmap_control.go`

File: pkg/controller/shoot/shoot_control.go [hunk 17]
```diff
@@ -32,13 +32,37 @@ import (
...
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		logger.Logger.Errorf("[ConfigMap watcher] Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+	c.reconcileConfigMapKey(key)
```
Comment (rfranzke) [inline pkg/controller/shoot/shoot_control.go]: This should be `c.configMapQueue.Add(key)`

File: pkg/controller/shoot/shoot_control.go [hunk 18]
```diff
@@ -32,13 +32,37 @@ import (
...
+}
+
+func (c *Controller) configMapUpdate(oldObj, newObj interface{}) {
+	var (
+		oldConfigMap = oldObj.(*corev1.ConfigMap)
+		newConfigMap = newObj.(*corev1.ConfigMap)
+	)
+
+	if apiequality.Semantic.Equalities.DeepEqual(oldConfigMap.Data, newConfigMap.Data) {
+		logger.Logger.Debugf("[ConfigMap watcher] No update of the `.data` field of cm %v/%v. Do not requeue the shoot", oldConfigMap.Namespace, oldConfigMap.Name)
```
Comment (rfranzke) [inline pkg/controller/shoot/shoot_control.go]: s/Do not requeue the Shoot/Do not requeue the ConfigMap`

File: pkg/controller/shoot/shoot_control.go [hunk 19]
```diff
@@ -80,6 +104,39 @@ func (c *Controller) shootDelete(obj interface{}) {
 	c.getShootQueue(obj).Add(key)
 }
 
+func (c *Controller) reconcileConfigMapKey(key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	logger.Logger.Infof("[ConfigMap Watcher] config map %v/%v was updated", namespace, name)
```
Comment (rfranzke) [inline pkg/controller/shoot/shoot_control.go]: Let's please not log this, we are watching all configmaps in the cluster and this would pollute the log.

File: pkg/controller/shoot/shoot_control.go [hunk 20]
```diff
@@ -80,6 +104,39 @@ func (c *Controller) shootDelete(obj interface{}) {
...
+
+	err = c.reconcileShootsReferringConfigMap(name, namespace)
```
Comment (rfranzke) [inline pkg/controller/shoot/shoot_control.go]: simplify to `return c.reconcileShootsReferringConfigMap(name, namespace)`

File: pkg/controller/shoot/shoot_control.go [hunk 21]
```diff
@@ -80,6 +104,39 @@ func (c *Controller) shootDelete(obj interface{}) {
...
+	}
+	for _, shoot := range shoots {
+		if shoot.Spec.Kubernetes.KubeAPIServer != nil &&
+			shoot.Spec.Kubernetes.KubeAPIServer.AuditConfig != nil &&
+			shoot.Spec.Kubernetes.KubeAPIServer.AuditConfig.AuditPolicy != nil &&
+			shoot.Spec.Kubernetes.KubeAPIServer.AuditConfig.AuditPolicy.Name == configMapName {
+			if shootKey, err := cache.MetaNamespaceKeyFunc(shoot); err == nil {
+				logger.Logger.Infof("[ConfigMap Watcher] schedule for reconciliation shoot %v/%v ", shoot.Namespace, shoot.Name)
+				c.shootQueue.Add(shootKey)
+			}
```
Comment (rfranzke) [inline pkg/controller/shoot/shoot_control.go]: Please log the error in case something goes wrong here.

File: pkg/operation/hybridbotanist/controlplane.go [hunk 22]
```diff
@@ -287,6 +287,27 @@ func (b *HybridBotanist) DeployKubeAPIServer() error {
 				admissionPlugins = append(admissionPlugins, plugin)
 			}
 		}
+
+		if apiServerConfig.AuditConfig != nil && apiServerConfig.AuditConfig.AuditPolicy != nil {
+
```
Comment (rfranzke) [inline pkg/operation/hybridbotanist/controlplane.go]: remove empty line here

File: pkg/operation/hybridbotanist/controlplane.go [hunk 23]
```diff
@@ -287,6 +287,27 @@ func (b *HybridBotanist) DeployKubeAPIServer() error {
...
+				return err
+			}
+			auditPolicy, ok := auditPolicyCm.Data["policy"]
+			if !ok {
+				return fmt.Errorf("Missing '.data.policy' in audit policy configmap %v/%v", cmRefNamespace, cmRefName)
+			}
+			if len(auditPolicy) == 0 {
+				return fmt.Errorf("Invalid audit policy. Provide non-empty audit policy")
+			}
+			// TODO: check provided policy is valid audit policy according to "k8s.io/apiserver/pkg/apis/audit[v1alpha1|v1beta1|v1]"
```
Comment (rfranzke) [inline pkg/operation/hybridbotanist/controlplane.go]: Cool, nice idea! What about this TODO? Can't we just call https://github.com/kubernetes/apiserver/blob/master/pkg/apis/audit/validation/validation.go#L28?

File: plugin/pkg/global/resourcereferencemanager/admission.go [hunk 24]
```diff
@@ -329,9 +336,26 @@ func (r *ReferenceManager) ensureShootReferences(shoot *garden.Shoot) error {
...
+		}
+	}
+
 	return nil
 }
 
+func isSpecifiedCustomAuditPolicy(apiServerConfig *garden.KubeAPIServerConfig) bool {
+	if apiServerConfig != nil && apiServerConfig.AuditConfig != nil {
+		if apiServerConfig.AuditConfig.AuditPolicy != nil {
+			if len(apiServerConfig.AuditConfig.AuditPolicy.Name) != 0 {
```
Comment (rfranzke) [inline plugin/pkg/global/resourcereferencemanager/admission.go]: Please merge the `if` conditions and combine them with `&&` (reduces too deeply nested expressions).

**Review by rfranzke — CHANGES_REQUESTED** [review]
Looks very good already, I've put a few suggestions inline. PTAL, thank you!

File: pkg/controller/shoot/shoot_control.go (hunk 19 above)
Comment (vpnachev) [reply pkg/controller/shoot/shoot_control.go]: What do you think about to switch it to `Debugf`?

Comment (rfranzke) [reply pkg/controller/shoot/shoot_control.go]: Not sure, we only care about the ConfigMaps that are referenced by shoots. Therefore, the log message in line 112 is totally fine, but this log message doesn't help at all, does it?

**Comment by vpnachev** [comment]
@rfranzke, this PR is ready for another round of review.
thanks!

**Comment by vpnachev** [comment]
I have updated the docu comments in the `pkg/apis/garden/[v1beta1/]types.go` files related to the custom audit policy. `pkg/openapi/openapi_generated.go` is also updated.

File: pkg/apis/garden/validation/validation_test.go [hunk 25]
```diff
@@ -4487,6 +4494,27 @@ var _ = Describe("validation", func() {
 			})
 		})
 
+		Context("AuditConfig validation", func() {
+			It("should forbid empty name", func() {
+				shoot.Spec.Kubernetes.KubeAPIServer.AuditConfig.AuditPolicy.ConfigMapRef.Name = ""
+				errorList := ValidateShoot(shoot)
+
+				Expect(len(errorList)).To(Equal(1))
+				Expect(*errorList[0]).To(MatchFields(IgnoreExtras, Fields{
```
Comment (adracus) [inline pkg/apis/garden/validation/validation_test.go]: ```Expect(errorList).To(ConsistOf(MatchFields...```

File: pkg/apis/garden/validation/validation_test.go [hunk 26]
```diff
@@ -4487,6 +4494,27 @@ var _ = Describe("validation", func() {
...
+					"Type":  Equal(field.ErrorTypeRequired),
+					"Field": Equal("spec.kubernetes.kubeAPIServer.auditConfig.auditPolicy.configMapRef.name"),
+				}))
+			})
+
+			It("should allow nil AuditConfig", func() {
+				shoot.Spec.Kubernetes.KubeAPIServer.AuditConfig = nil
+				errorList := ValidateShoot(shoot)
+
+				Expect(len(errorList)).To(Equal(0))
```
Comment (adracus) [inline pkg/apis/garden/validation/validation_test.go]: `Expect(errorList).To(BeEmpty())`

**Review by rfranzke — APPROVED** [review]
/lgtm
